db/constraints.c

```diff
@@ -2141,97 +2141,160 @@ static inline int key_has_expressions_members(struct schema *key)
+    }
 
-            /* If we have a target table (to_db) only look at rules pointing
-             * to that table. */
-            if (to_db && strcasecmp(ct->table[jj], to_db->tablename) != 0)
-                continue;
+    if (constraint_key_check(fky, bky)) {
+        constraint_err(s, source_db, ct, rule_ix, "invalid number of columns");
+        n_errors++;
+    }
 
-            rdb = get_dbtable_by_name(ct->table[jj]);
-            if (rdb)
-                rdb = get_newer_db(rdb, new_db);
-            else if (strcasecmp(ct->table[jj], from_db->tablename) == 0)
-                rdb = from_db;
-            if (!rdb) {
-                /* Referencing a non-existent table */
-                constraint_err(s, from_db, ct, jj, "parent table not found");
-                n_errors++;
-                continue;
-            } else {
-                if (rdb->timepartition_name) {
-                    constraint_err(s, from_db, ct, jj, "A foreign key cannot refer to a time partition");
-                    n_errors++;
-                    continue;
-                }
-            }
-            if (rdb == new_db) {
-                snprintf(keytag, sizeof(keytag), ".NEW.%s", ct->keynm[jj]);
-                if (!(bky = find_tag_schema_by_name(ct->table[jj], keytag))) {
-                    /* Referencing a nonexistent key */
-                    constraint_err(s, from_db, ct, jj, "parent key not found");
-                    n_errors++;
-                }
-            } else {
-                snprintf(keytag, sizeof(keytag), "%s", ct->keynm[jj]);
-                if (!(bky = find_tag_schema(rdb, keytag))) {
-                    /* Referencing a nonexistent key */
-                    constraint_err(s, from_db, ct, jj, "parent key not found");
-                    n_errors++;
-                }
-            }
+    return n_errors;
+}
 
-            if (constraint_key_check(fky, bky)) {
-                /* Invalid constraint index */
-                constraint_err(s, from_db, ct, jj, "invalid number of columns");
-                n_errors++;
-            }
+/*
+ * Helper function for `verify_constraints_from_dbtable`/`verify_constraints_to_dbtable`
+ *
+ * Performs the following verifications
+ *     1. `source_db` has the constrained-on key
+ *     2. Constraint does not have a cascading update on expression indexes
+ *     3. Each of the constraint's rules is valid. See `verify_constraint_rule` for
+ *       details on rule verification. (if `target_db` is provided,
+ *       only rules referring to `target_db` are checked)
+ *
+ * `source_db_is_updated`/`target_db_is_updated` should be 1 if 
+ * `source_db`/`target_db` refers to the new btree for an existing table.
+ *
+ * Logs each failure and returns the number of failed verifications.
+ */
+static int verify_constraint(constraint_t * const ct,
+                             struct dbtable *source_db,
+                             struct dbtable *target_db,
```

Comment:
TODO: Give `source_db` and `target_db` variables clearer names--maybe `ct_source_db`/`ct_target_db`?